WCF/Shared/Implementation/ClientTelemetryChannelBase.cs

```diff
@@ -0,0 +1,361 @@
+        {
+            if ( result == null )
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
+            var oar = (OpenAsyncResult)result;
+            StopOpenTelemetry(oar.Telemetry, oar.LastException, nameof(OpenCompleted));
+        }
+
+        public IAsyncResult BeginClose(AsyncCallback callback, object state)
+        {
+            return BeginClose(ChannelManager.CloseTimeout, callback, state);
+        }
+
+        public IAsyncResult BeginClose(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            WcfClientEventSource.Log.ChannelCalled(GetType().FullName, nameof(BeginClose));
+            return InnerChannel.BeginClose(timeout, callback, state);
+        }
+
+        public void EndClose(IAsyncResult result)
+        {
+            if ( result == null )
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
+            WcfClientEventSource.Log.ChannelCalled(GetType().FullName, nameof(EndClose));
+            try
+            {
+                InnerChannel.EndClose(result);
+            } finally
+            {
+                Dispose(true);
+            }
+        }
+
+
+        private void HookChannelEvents()
+        {
+            InnerChannel.Closed += OnChannelClosed;
+            InnerChannel.Closing += OnChannelClosing;
+            InnerChannel.Opened += OnChannelOpened;
+            InnerChannel.Opening += OnChannelOpening;
+            InnerChannel.Faulted += OnChannelFaulted;
+        }
+        private void UnhookChannelEvents()
+        {
+            InnerChannel.Closed -= OnChannelClosed;
+            InnerChannel.Closing -= OnChannelClosing;
+            InnerChannel.Opened -= OnChannelOpened;
+            InnerChannel.Opening -= OnChannelOpening;
+            InnerChannel.Faulted -= OnChannelFaulted;
+        }
+        private void OnChannelOpened(object sender, EventArgs e)
+        {
+            Opened?.Invoke(this, e);
+        }
+        private void OnChannelOpening(object sender, EventArgs e)
+        {
+            Opening?.Invoke(this, e);
+        }
+        private void OnChannelClosed(object sender, EventArgs e)
+        {
+            Closed?.Invoke(sender, e);
+        }
+        private void OnChannelClosing(object sender, EventArgs e)
+        {
+            Closing?.Invoke(sender, e);
+        }
+        private void OnChannelFaulted(object sender, EventArgs e)
+        {
+            Faulted?.Invoke(sender, e);
+        }
+
+
+        // telemetry implementation
+        private DependencyTelemetry StartOpenTelemetry(String method)
+        {
+            try
+            {
+                var telemetry = new DependencyTelemetry();
+                telemetry.Start();
+                telemetry.Type = DependencyConstants.WcfChannelOpen;
+                telemetry.Target = RemoteAddress.Uri.Host;
+                telemetry.Name = RemoteAddress.Uri.ToString();
```

Comment:
Name should have small cardinality. `Uri` likely will have arguments that should not be a part of name.